skrub/_interpolation_join.py

```diff
@@ -0,0 +1,343 @@
+import joblib
+import pandas as pd
+from sklearn.base import BaseEstimator, TransformerMixin, clone
+from sklearn.ensemble import (
+    HistGradientBoostingClassifier,
+    HistGradientBoostingRegressor,
+)
+from sklearn.utils._tags import _safe_tags
+
+from skrub import _utils
+from skrub._table_vectorizer import TableVectorizer
+
+
+class InterpolationJoiner(TransformerMixin, BaseEstimator):
+    """Join with a table augmented by machine-learning predictions.
+
+    This is similar to a usual equi-join, but instead of looking for actual
+    rows in the right table that satisfy the join condition, we estimate what
+    those rows would contain if they existed in the table.
+
+    Suppose we want to join a table ``buildings(latitude, longitude, n_stories)``
+    with a table ``annual_avg_temp(latitude, longitude, avg_temp)``. Our annual
+    average temperature table may not contain data for the exact latitude and
+    longitude of our buildings. However, we can interpolate what we need from
+    the data points it does contain. Using ``annual_avg_temp``, we train a
+    model to predict the temperature, given the latitude and longitude. Then,
+    we use this model to estimate the values we want to add to our
+    ``buildings`` table. In a way we are joining ``buildings`` to a virtual
+    table, in which rows for any (latitude, longitude) location are inferred,
+    rather than retrieved, when requested. This is done with::
+
+        InterpolationJoiner(
+            annual_avg_temp, on=["latitude", "longitude"]
+        ).fit_transform(buildings)
```

Comment:
This is somewhat linked to what you do in the simple reproducible example below. Why not just add these explanations below?

Reply:
I thought having the explanations sooner would help understand the list of parameters which comes just below. the Examples section appears towards the end (both in the docstring and the rendered doc) and ideally we would already have a good understanding of what the function does before reaching the detailed parameter list. But I can move that paragraph if you think it belongs in the examples section

Reply:
I agree that we need a paragraph about the mains parameters here, but I find this part redundant with the example below. 
Also, I don't find the simple example is towards the end (it comes before methods and attributes) and I often look at it, for instance in the pandas or polars docs, just after having a glimpse at the parameters.
How about:
```We have a main and auxiliary table that we join on imperfect correspondences. Instead of matching on the nearest neighbor in the auxiliary table, the objective of the InterpolationJoiner is to infer and estimate from close data points the corresponding value(s). This is particularly useful in temporal or spatial joins (see example with latitude and longitude below).```
and incorporating your text into the example?

Reply:
This is a way of documenting a class I've found clearer than what I'm used to. There is a slight redundancy, but it doesn't hurt the user experience, IMHO. I especially appreciate using the example in the parameter section as well.
What do the other @skrub-data/devs think?